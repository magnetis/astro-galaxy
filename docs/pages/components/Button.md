# Button

Buttons define what action will occur when the user interacts with features and components of the platform. Buttons are used to initialize and dictate the actions on the interface, either in the background or foreground of an experience.

## primary buttons

Primary buttons are medium sized and have a colored background. They aren't always the same color, though, so there isn't a default "primary" look or CSS class - primary here is more of a concept.

Please note that color and size are built through CSS classes applied to the `<button>` element. You'll always need to specify color and size to get a complete button.

Use these for the main action in a section or screen. You should display only one primary button at a time.

Mouseover and click to view :hover and :active states.

```.jsx
<Button>Button</Button>
```
