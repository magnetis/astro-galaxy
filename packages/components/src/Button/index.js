import styled from 'styled-components';
import { color, space, border, layout, fontSize, buttonStyle, typography } from 'styled-system';
import { font, buttonSizes } from '@magnetis/astro-galaxy-core';

const hoverStyles = props => ({
  'primary.uranus': `
    background-color: ${props.theme.colors.uranus300};
    border-color: ${props.theme.colors.uranus300};
  `,
  'primary.earth': `
    background-color: ${props.theme.colors.earth200};
    border-color: ${props.theme.colors.earth200};
  `,
  'primary.venus': `
    background-color: ${props.theme.colors.venus200};
    border-color: ${props.theme.colors.venus200};
  `,
  'primary.mars': `
    background-color: ${props.theme.colors.mars300};
    border-color: ${props.theme.colors.mars300};
  `,
});

const activeStyles = props => ({
  'primary.uranus': `
    background-color: ${props.theme.colors.uranus600};
    border-color: ${props.theme.colors.uranus600};
  `,
  'primary.earth': `
    background-color: ${props.theme.colors.earth500};
    border-color: ${props.theme.colors.earth500};
  `,
  'primary.venus': `
    background-color: ${props.theme.colors.venus500};
    border-color: ${props.theme.colors.venus500};
  `,
  'primary.mars': `
    background-color: ${props.theme.colors.mars600};
    border-color: ${props.theme.colors.mars600};
  `,
});

const Button = styled.button`
  cursor: pointer;
  white-space: nowrap;
  text-align: center;
  transition: all 0.3s ease;
  overflow: hidden;
  border-width: 2px;
  border-style: solid;

  :disabled,
  .disabled {
    cursor: not-allowed;
    background-color: ${props => props.theme.colors.moon200};
    border-color: ${props => props.theme.colors.moon200};
    color: ${props => props.theme.colors.space100};
  }

  :hover:not(:disabled):not(.disabled),
  :focus:not(:disabled):not(.disabled) {
    ${props => hoverStyles(props)[props.variant]}
  }

  :active:not(:disabled):not(.disabled) {
    ${props => activeStyles(props)[props.variant]}
  }

  :focus {
    outline: none;
  }

  ${font};
  ${color};
  ${space};
  ${border};
  ${layout};
  ${fontSize};
  ${typography};
  ${buttonStyle};
  ${buttonSizes};
`;

Button.defaultProps = {
  display: 'inline-block',
  variant: 'primary.uranus',
  buttonSize: 'medium',
  font: 'primary',
};

Button.propTypes = {
  ...color.propTypes,
  ...space.propTypes,
  ...border.propTypes,
  ...layout.propTypes,
  ...fontSize.propTypes,
  ...buttonStyle.propTypes,
  ...typography.propTypes,
};

export default Button;
