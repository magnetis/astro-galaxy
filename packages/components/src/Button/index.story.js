import React from 'react';
import { storiesOf } from '@storybook/react';
import Button from './index.js';

storiesOf('buttons', module)
  .add('button default', () => <Button>uranus</Button>)
  .add('primary buttons variants', () => (
    <React.Fragment>
      <Button variant="primary.uranus">uranus</Button>
      <Button variant="primary.earth">earth</Button>
      <Button variant="primary.venus">venus</Button>
      <Button variant="primary.mars">mars</Button>
      <Button disabled>disabled</Button>
    </React.Fragment>
  ));

storiesOf('button sizes', module).add('primary buttons sizes', () => (
  <React.Fragment>
    <Button buttonSize="small">small</Button>
    <Button buttonSize="medium">medium ~ default</Button>
    <Button buttonSize="large">large</Button>
  </React.Fragment>
));
