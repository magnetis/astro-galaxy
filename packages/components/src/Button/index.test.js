import React from 'react';
import toJSON from 'enzyme-to-json';
import { mountWithTheme } from '@magnetis/astro-galaxy-core';
import Button from './index';

describe('Button', () => {
  it('should render a Button component', () => {
    const wrapper = mountWithTheme(<Button>uranus</Button>);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
