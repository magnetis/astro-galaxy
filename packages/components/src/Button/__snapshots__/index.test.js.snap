// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render a Button component 1`] = `
<Button
  buttonSize="medium"
  display="inline-block"
  font="primary"
  variant="primary.uranus"
>
  <StyledComponent
    buttonSize="medium"
    display="inline-block"
    font="primary"
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "_foldedDefaultProps": Object {
          "buttonSize": "medium",
          "display": "inline-block",
          "font": "primary",
          "variant": "primary.uranus",
        },
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Button-tgj28e-0",
          "isStatic": false,
          "lastClassName": "eIrlVm",
          "rules": Array [
            "cursor:pointer;white-space:no-wrap;text-align:center;transition:all 0.3s ease;overflow:hidden;border-width:2px;border-style:solid;:disabled{cursor:not-allowed;background-color:",
            [Function],
            ";border-color:",
            [Function],
            ";color:",
            [Function],
            ";}:hover:not(:disabled),:focus:not(:disabled){",
            [Function],
            "}:active:not(:disabled){",
            [Function],
            "}:focus{outline:none;}",
            [Function],
            ";",
            [Function],
            ";",
            [Function],
            ";",
            [Function],
            ";",
            [Function],
            ";",
            [Function],
            ";",
            [Function],
            ";",
            [Function],
            ";",
            [Function],
            ";",
          ],
        },
        "displayName": "Button",
        "foldedComponentIds": Array [],
        "propTypes": Object {},
        "render": [Function],
        "styledComponentId": "Button-tgj28e-0",
        "target": "button",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    variant="primary.uranus"
  >
    <button
      className="Button-tgj28e-0 eIrlVm"
      display="inline-block"
    >
      uranus
    </button>
  </StyledComponent>
</Button>
`;
